File: backend/fetcher.js
const https = require('https');
const http = require('http');

/**
 * Helper to fetch text or JSON over HTTPS with timeout
 */
function fetchUrl(url, timeoutMs = 12000) {
  return new Promise((resolve, reject) => {
    const client = url.startsWith('https') ? https : http;
    const req = client.get(url, { headers: { 'User-Agent': 'katproxy/1.0' } }, (res) => {
      if (res.statusCode < 200 || res.statusCode >= 300) {
        res.resume();
        return reject(new Error(`Failed to fetch ${url}, status: ${res.statusCode}`));
      }
      let rawData = '';
      res.setEncoding('utf8');
      res.on('data', (chunk) => { rawData += chunk; });
      res.on('end', () => resolve(rawData));
    });

    req.on('error', (err) => reject(err));
    req.setTimeout(timeoutMs, () => {
      req.destroy();
      reject(new Error(`Timeout fetching ${url}`));
    });
  });
}

/**
 * Fetch and merge proxies from multiple premier GitHub repositories
 */
async function fetchAllProxies() {
  console.log('[Fetcher] Starting proxy list aggregation...');
  const proxyMap = new Map();

  // Helper to add or merge a proxy
  function addProxy(host, port, protocol, extra = {}) {
    if (!host || !port) return;
    const cleanHost = String(host).trim();
    const cleanPort = parseInt(port, 10);
    if (isNaN(cleanPort) || cleanPort <= 0 || cleanPort > 65535) return;
    // Basic IP validation regex
    if (!/^(\d{1,3}\.){3}\d{1,3}$/.test(cleanHost)) return;

    const key = `${cleanHost}:${cleanPort}`;
    const existing = proxyMap.get(key);

    const protoClean = (protocol || 'http').toLowerCase();

    if (!existing) {
      proxyMap.set(key, {
        id: key,
        host: cleanHost,
        port: cleanPort,
        protocols: [protoClean],
        country: (extra.country || 'UN').toUpperCase(),
        countryName: extra.countryName || 'Unknown',
        city: extra.city || '',
        org: extra.org || '',
        latency: extra.latency != null ? Math.round(extra.latency * 1000) : null,
        alive: extra.alive !== undefined ? extra.alive : null,
        lastChecked: extra.lastChecked || null,
        sources: [extra.source || 'upstream']
      });
    } else {
      if (!existing.protocols.includes(protoClean)) {
        existing.protocols.push(protoClean);
      }
      if (existing.country === 'UN' && extra.country) {
        existing.country = extra.country.toUpperCase();
        existing.countryName = extra.countryName || existing.countryName;
      }
      if (!existing.org && extra.org) {
        existing.org = extra.org;
      }
      if (extra.source && !existing.sources.includes(extra.source)) {
        existing.sources.push(extra.source);
      }
    }
  }

  // 1. Fetch monosans structured JSON (provides rich metadata: country, org, timeout)
  try {
    console.log('[Fetcher] Fetching monosans/proxy-list JSON...');
    const monosansData = await fetchUrl('https://raw.githubusercontent.com/monosans/proxy-list/main/proxies.json', 15000);
    const parsed = JSON.parse(monosansData);
    if (Array.isArray(parsed)) {
      for (const item of parsed) {
        const countryCode = item.geolocation?.country?.iso_code || 'UN';
        const countryName = item.geolocation?.country?.names?.en || 'Unknown';
        const city = item.geolocation?.city?.names?.en || '';
        const org = item.asn?.autonomous_system_organization || '';
        const latencyMs = item.timeout ? Math.round(item.timeout * 1000) : null;

        addProxy(item.host, item.port, item.protocol, {
          country: countryCode,
          countryName,
          city,
          org,
          latency: item.timeout != null ? item.timeout : null,
          alive: true,
          source: 'monosans'
        });
      }
      console.log(`[Fetcher] monosans parsed: ${parsed.length} proxies`);
    }
  } catch (err) {
    console.warn('[Fetcher] Warning: monosans JSON fetch failed:', err.message);
  }

  // 2. TheSpeedX lists (HTTP, SOCKS4, SOCKS5)
  const speedXTasks = [
    { url: 'https://raw.githubusercontent.com/TheSpeedX/PROXY-List/master/http.txt', protocol: 'http' },
    { url: 'https://raw.githubusercontent.com/TheSpeedX/PROXY-List/master/socks4.txt', protocol: 'socks4' },
    { url: 'https://raw.githubusercontent.com/TheSpeedX/PROXY-List/master/socks5.txt', protocol: 'socks5' }
  ];

  await Promise.allSettled(
    speedXTasks.map(async (task) => {
      try {
        const text = await fetchUrl(task.url);
        const lines = text.split('\n');
        let count = 0;
        for (const line of lines) {
          const trimmed = line.trim();
          if (!trimmed || trimmed.startsWith('#')) continue;
          const [host, port] = trimmed.split(':');
          if (host && port) {
            addProxy(host, port, task.protocol, { source: 'thespeedx' });
            count++;
          }
        }
        console.log(`[Fetcher] TheSpeedX ${task.protocol} parsed: ${count} proxies`);
      } catch (err) {
        console.warn(`[Fetcher] TheSpeedX ${task.protocol} error:`, err.message);
      }
    })
  );

  // 3. hookzof SOCKS5 list
  try {
    const hookzofText = await fetchUrl('https://raw.githubusercontent.com/hookzof/socks5_list/master/proxy.txt');
    const lines = hookzofText.split('\n');
    let count = 0;
    for (const line of lines) {
      const trimmed = line.trim();
      if (!trimmed || trimmed.startsWith('#')) continue;
      const [host, port] = trimmed.split(':');
      if (host && port) {
        addProxy(host, port, 'socks5', { source: 'hookzof' });
        count++;
      }
    }
    console.log(`[Fetcher] hookzof SOCKS5 parsed: ${count} proxies`);
  } catch (err) {
    console.warn('[Fetcher] hookzof SOCKS5 error:', err.message);
  }

  const results = Array.from(proxyMap.values());
  console.log(`[Fetcher] Completed aggregation. Total unique proxies: ${results.length}`);
  return results;
}

module.exports = {
  fetchAllProxies
};

File: backend/src/fetcher.rs
use std::collections::HashMap;
use std::time::Duration;
use reqwest::Client;
use tracing::{info, warn};
use crate::models::{MonosansProxy, Proxy};

pub async fn fetch_all_proxies(client: &Client) -> Vec<Proxy> {
    info!("[Fetcher] Starting proxy list aggregation...");
    let mut proxy_map: HashMap<(String, u16), Proxy> = HashMap::new();

    // Helper closure to insert or merge
    let mut add_proxy = |host: String, port: u16, protocol: String, country: String, country_name: String, city: String, org: String, latency: Option<u32>, alive: Option<bool>, source: String| {
        let clean_host = host.trim().to_string();
        if clean_host.is_empty() || port == 0 {
            return;
        }

        let key = (clean_host.clone(), port);
        let proto_clean = protocol.to_lowercase();

        if let Some(existing) = proxy_map.get_mut(&key) {
            if !existing.protocols.contains(&proto_clean) {
                existing.protocols.push(proto_clean);
            }
            if existing.country == "UN" && country != "UN" {
                existing.country = country;
                if !country_name.is_empty() {
                    existing.country_name = country_name;
                }
            }
            if existing.org.is_empty() && !org.is_empty() {
                existing.org = org;
            }
            if !existing.sources.contains(&source) {
                existing.sources.push(source);
            }
        } else {
            let id = format!("{}:{}", clean_host, port);
            proxy_map.insert(key, Proxy {
                id,
                host: clean_host,
                port,
                protocols: vec![proto_clean],
                country,
                country_name,
                city,
                org,
                latency,
                alive,
                last_checked: None,
                sources: vec![source],
            });
        }
    };

    // 1. Fetch monosans structured JSON
    let monosans_url = "https://raw.githubusercontent.com/monosans/proxy-list/main/proxies.json";
    match client.get(monosans_url).timeout(Duration::from_secs(15)).send().await {
        Ok(resp) => {
            if resp.status().is_success() {
                if let Ok(items) = resp.json::<Vec<MonosansProxy>>().await {
                    let count = items.len();
                    for item in items {
                        let country = item.geolocation.as_ref()
                            .and_then(|g| g.country.as_ref())
                            .and_then(|c| c.iso_code.clone())
                            .unwrap_or_else(|| "UN".to_string())
                            .to_uppercase();

                        let country_name = item.geolocation.as_ref()
                            .and_then(|g| g.country.as_ref())
                            .and_then(|c| c.names.as_ref())
                            .and_then(|n| n.get("en").cloned())
                            .unwrap_or_else(|| "Unknown".to_string());

                        let city = item.geolocation.as_ref()
                            .and_then(|g| g.city.as_ref())
                            .and_then(|c| c.names.as_ref())
                            .and_then(|n| n.get("en").cloned())
                            .unwrap_or_default();

                        let org = item.asn.as_ref()
                            .and_then(|a| a.autonomous_system_organization.clone())
                            .unwrap_or_default();

                        let latency = item.timeout.map(|t| (t * 1000.0).round() as u32);

                        add_proxy(
                            item.host,
                            item.port,
                            item.protocol,
                            country,
                            country_name,
                            city,
                            org,
                            latency,
                            Some(true),
                            "monosans".to_string(),
                        );
                    }
                    info!("[Fetcher] monosans parsed: {} proxies", count);
                }
            }
        }
        Err(e) => warn!("[Fetcher] monosans fetch error: {}", e),
    }

    // 2. TheSpeedX lists (HTTP, SOCKS4, SOCKS5)
    let speedx_tasks = vec![
        ("https://raw.githubusercontent.com/TheSpeedX/PROXY-List/master/http.txt", "http"),
        ("https://raw.githubusercontent.com/TheSpeedX/PROXY-List/master/socks4.txt", "socks4"),
        ("https://raw.githubusercontent.com/TheSpeedX/PROXY-List/master/socks5.txt", "socks5"),
    ];

    for (url, proto) in speedx_tasks {
        match client.get(url).timeout(Duration::from_secs(12)).send().await {
            Ok(resp) => {
                if resp.status().is_success() {
                    if let Ok(text) = resp.text().await {
                        let mut count = 0;
                        for line in text.lines() {
                            let trimmed = line.trim();
                            if trimmed.is_empty() || trimmed.starts_with('#') {
                                continue;
                            }
                            if let Some((h, p)) = trimmed.split_once(':') {
                                if let Ok(port) = p.parse::<u16>() {
                                    add_proxy(
                                        h.to_string(),
                                        port,
                                        proto.to_string(),
                                        "UN".to_string(),
                                        "Unknown".to_string(),
                                        String::new(),
                                        String::new(),
                                        None,
                                        None,
                                        "thespeedx".to_string(),
                                    );
                                    count += 1;
                                }
                            }
                        }
                        info!("[Fetcher] TheSpeedX {} parsed: {} proxies", proto, count);
                    }
                }
            }
            Err(e) => warn!("[Fetcher] TheSpeedX {} error: {}", proto, e),
        }
    }

    // 3. hookzof SOCKS5 list
    let hookzof_url = "https://raw.githubusercontent.com/hookzof/socks5_list/master/proxy.txt";
    match client.get(hookzof_url).timeout(Duration::from_secs(12)).send().await {
        Ok(resp) => {
            if resp.status().is_success() {
                if let Ok(text) = resp.text().await {
                    let mut count = 0;
                    for line in text.lines() {
                        let trimmed = line.trim();
                        if trimmed.is_empty() || trimmed.starts_with('#') {
                            continue;
                        }
                        if let Some((h, p)) = trimmed.split_once(':') {
                            if let Ok(port) = p.parse::<u16>() {
                                add_proxy(
                                    h.to_string(),
                                    port,
                                    "socks5".to_string(),
                                    "UN".to_string(),
                                    "Unknown".to_string(),
                                    String::new(),
                                    String::new(),
                                    None,
                                    None,
                                    "hookzof".to_string(),
                                );
                                count += 1;
                            }
                        }
                    }
                    info!("[Fetcher] hookzof SOCKS5 parsed: {} proxies", count);
                }
            }
        }
        Err(e) => warn!("[Fetcher] hookzof error: {}", e),
    }

    let results: Vec<Proxy> = proxy_map.into_values().collect();
    info!("[Fetcher] Completed aggregation. Total unique proxies: {}", results.len());
    results
}
